### PR TITLE
Performance improvement in libcudf case conversion for long strings

### LINKED ISSUE #15406: [FEA] lower/upper are really slow for long strings
**Is your feature request related to a problem? Please describe.**

We have a customer that is doing `lower` on strings that are rather long. I am not 100% sure if they include multi-byte characters or not, but there are at least other data sets from this customer that do include multi-byte characters.  One of the issues we are running into is that the performance for `lower` is really bad when there are long strings, which is the case for this customer.

I ran a simple benchmark using SparkRapids on an RTX a6000 GPU.

 |avg Size (bytes) | total size GiB | baseline (ms) | with lower (ms) | diff (ms) | Throughput GB/s|
-- | -- | -- | -- | -- | --
29.89 | 10.01 | 2,153 | 2,977 | 824 | 12.15
130.99 | 10.01 | 1,881 | 4,104 | 2,223 | 4.50
962.96 | 10.01 | 1,751 | 4,748 | 2,997 | 3.34
7,820.52 | 10.00 | 1,688 | 3,898 | 2,210 | 4.53

The baseline is just reading in the data from a parquet file and doing a SUM on the octet_length of the string. (that is why it actually gets faster when the length of the string increases, because it is not processing as much data after that). "with lower" is the same thing, but first calls `lower` on the column before doing the `octet_length`. I did this 5 times for each query and the median value is shown.

In theory the GPU should have 768 GB/s memory bandwidth. 12 GiB/s is not great, but is a lot better than the 3 or 4 we are seeing for the longer strings.

**Describe the solution you'd like**
I would like it to be faster. Perhaps we can have a long string version for operations with longer strings. Perhaps we can optimize the current version to load the data in a better way too. Not really sure.

**Describe alternatives you've considered**
We are looking at ways to call this kernel less often (bigger batches), but it kind of just kicks the can down the road.

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Improves logic efficiency overall strings case conversion and reworks the specialized kernels for long strings to improve parallelization within each string.
Closes #15406 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
Existing benchmark show good improvement for long strings
```
## [0] NVIDIA RTX A6000

| row_width | num_rows |   Ref Time |   Cmp Time |           Diff |   %Diff |
|-----------|----------|------------|------------|----------------|---------|
|    32     |   4096   | 105.801 us |  98.869 us |      -6.932 us |  -6.55% |
|    64     |   4096   | 142.955 us | 138.110 us |      -4.845 us |  -3.39% |
|    128    |   4096   | 186.503 us | 211.072 us |      24.569 us |  13.17% |
|    256    |   4096   | 288.387 us | 214.724 us |     -73.663 us | -25.54% |
|    512    |   4096   | 514.895 us | 224.731 us |    -290.164 us | -56.35% |
|   1024    |   4096   |   1.007 ms | 262.894 us |    -743.903 us | -73.89% |
|   2048    |   4096   |   1.992 ms | 348.787 us |   -1643.051 us | -82.49% |
|    32     |  32768   | 109.572 us | 102.730 us |      -6.843 us |  -6.24% |
|    64     |  32768   | 147.685 us | 143.177 us |      -4.508 us |  -3.05% |
|    128    |  32768   | 237.028 us | 237.712 us |       0.684 us |   0.29% |
|    256    |  32768   | 354.450 us | 340.402 us |     -14.048 us |  -3.96% |
|    512    |  32768   | 651.191 us | 529.697 us |    -121.494 us | -18.66% |
|   1024    |  32768   |   1.365 ms | 914.432 us |    -450.780 us | -33.02% |
|   2048    |  32768   |   2.668 ms |   1.707 ms |    -961.703 us | -36.04% |
|    32     |  262144  | 192.874 us | 174.467 us |     -18.406 us |  -9.54% |
|    64     |  262144  | 401.702 us | 409.239 us |       7.538 us |   1.88% |
|    128    |  262144  |   2.059 ms | 961.064 us |   -1097.503 us | -53.31% |
|    256    |  262144  |   6.995 ms |   1.587 ms |   -5407.260 us | -77.31% |
|    512    |  262144  |  18.476 ms |   2.916 ms |  -15560.385 us | -84.22% |
|   1024    |  262144  |  41.705 ms |   6.044 ms |  -35661.092 us | -85.51% |
|   2048    |  262144  |  98.148 ms |  12.489 ms |  -85658.725 us | -87.28% |
|    32     | 2097152  | 928.088 us | 762.243 us |    -165.845 us | -17.87% |
|    64     | 2097152  |   2.161 ms |   2.022 ms |    -138.884 us |  -6.43% |
|    128    | 2097152  |  14.398 ms |   6.582 ms |   -7816.360 us | -54.29% |
|    256    | 2097152  |  55.781 ms |  11.553 ms |  -44228.492 us | -79.29% |
|    512    | 2097152  | 150.897 ms |  22.285 ms | -128611.893 us | -85.23% |
|    32     | 16777216 |   7.042 ms |   5.710 ms |   -1331.347 us | -18.91% |
|    64     | 16777216 |  16.597 ms |  15.268 ms |   -1329.738 us |  -8.01% |

```

File: cpp/src/strings/case.cu [hunk 1]
```diff
@@ -137,45 +140,116 @@ struct upper_lower_fn {
...
+
+  size_type size = 0;
+  for (auto i = lane_idx; i < d_str.size_bytes(); i += cudf::detail::warp_size) {
+    auto const chr = str_ptr[i];
+    if (is_utf8_continuation_char(chr)) { continue; }
+    char_utf8 u8 = 0;
+    to_char_utf8(str_ptr + i, u8);
+    size += converter.process_character(u8);
+  }
+  // this is every so slightly faster than using the cub::warp_reduce
```
Comment (PointKernel) [inline cpp/src/strings/case.cu:246]: probably due to the high register usage with cub

File: cpp/src/strings/case.cu [hunk 2]
```diff
@@ -137,45 +140,116 @@ struct upper_lower_fn {
...
+    process_string(d_str, idx);
   }
 };
 
+/**
+ * @brief Count output bytes in warp-parallel threads
+ *
+ * This executes as one warp per string and just computes the output sizes.
+ */
+CUDF_KERNEL void count_bytes_fn(convert_char_fn converter,
```
Comment (PointKernel) [inline cpp/src/strings/case.cu:220]:
```suggestion
CUDF_KERNEL void count_bytes_kernel(convert_char_fn converter,
```
nit: `_fn` made me think it's a functor in the first place.

File: cpp/src/strings/case.cu [hunk 3]
```diff
@@ -235,40 +316,55 @@ std::unique_ptr<column> convert_case(strings_column_view const& input,
...
+  }();
+
+  // build sub-offsets
+  auto const input_chars = input.chars_begin(stream);
+  auto const sub_count   = chars_size / LS_SUB_BLOCK_SIZE;
+  auto tmp_offsets       = rmm::device_uvector<int64_t>(sub_count + input.size() + 1, stream);
+  {
+    rmm::device_uvector<size_type> sub_offsets(sub_count, stream);
+    auto const count_itr = thrust::make_counting_iterator<size_type>(0);
+    thrust::transform(rmm::exec_policy(stream),
```
Comment (PointKernel) [inline cpp/src/strings/case.cu:346]: question: could this algorithm use `exec_policy_nosync` policy as well?

**Review by ttnghia — APPROVED** [review]

**Review by PointKernel — APPROVED** [review]
Clean as always. LGTM

**Pull request merged**